### PR TITLE
Add Shutdown and Close Functions for Graceful Connection Termination

### PR DESCRIPTION
**Pull request by maohieng**
## Pull Request Template

**Description:**

This pull request introduces essential shutdown and close functions to our web services and connections, ensuring graceful termination of connections and releasing associated resources. This addition empowers users to implement their own graceful shutdown procedures as needed.
Closes: #663 

**Breaking Changes (if applicable):**

-   No breaking changes to existing functionalities.

**Additional Information:**

- **Shutdown Functionality**: Implement shutdown functions for web services and connections to gracefully terminate connections and release associated resources.
- **Close Function**: Add a close function to properly close connections and release resources in a controlled manner.
- **User Flexibility**: Enable users to implement custom graceful shutdown procedures tailored to their specific requirements.

**Checklist:**

-   [ ] I have formatted my code using  `goimport`  and  `golangci-lint`.
-   [ ] All new code is covered by unit tests.
-   [ ] This PR does not decrease the overall code coverage.
-   [ ] I have reviewed the code comments and documentation for clarity.

**Thank you for your contribution!**

### CODE REVIEW

File: pkg/gofr/datasource/redis/redis.go [hunk 1]
```diff
@@ -63,6 +63,15 @@ func NewClient(c config.Config, logger datasource.Logger, metrics Metrics) *Redi
 	return &Redis{Client: rc, config: redisConfig, logger: logger}
 }
 
+// Close shutdowns and save the state of the redis client.
```
Comment (ccoVeille) [inline pkg/gofr/datasource/redis/redis.go:66]: This block is added but commented

So it looks like a WIP/TODO

Is that normal?

File: pkg/gofr/gofr.go [hunk 2]
```diff
@@ -177,6 +178,29 @@ func (a *App) Run() {
...
+		err = errors.Join(err, a.grpcServer.Shutdown(ctx))
+	}
+
+	if a.metricServer != nil {
+		err = errors.Join(err, a.metricServer.Shutdown(ctx))
+	}
+
+	err = errors.Join(err, a.container.Close())
+
+	a.container.Logger.Infof("Application shutdown complete: %v", err)
```
Comment (ccoVeille) [inline pkg/gofr/gofr.go:199]: So the log with be always present 

But the message will be strange with the trailing ": ' when there is no error, right?

And if an error occurs, it should be reported as a warning/error, no?

File: pkg/gofr/datasource/pubsub/mqtt/mqtt.go [hunk 3]
```diff
@@ -322,6 +322,11 @@ func (m *MQTT) Unsubscribe(topic string) error {
 	return nil
 }
 
+func (m *MQTT) Close() error {
+	m.Client.Disconnect(250)
```
Comment (ccoVeille) [inline pkg/gofr/datasource/pubsub/mqtt/mqtt.go:326]: 250 is a magic number here

Please add a variable or a const and a comment about its meaning

File: pkg/gofr/datasource/pubsub/google/google_test.go [hunk 4]
```diff
@@ -50,6 +50,24 @@ func TestGoogleClient_New_Error(t *testing.T) {
...
+	defer ctrl.Finish()
+
+	logger := logging.NewMockLogger(logging.ERROR)
+	g = New(Config{}, logger, NewMockMetrics(ctrl))
+
+	assert.Nil(t, g)
+
+	err := g.Close()
+
+	assert.Nil(t, err)
```
Comment (ccoVeille) [inline pkg/gofr/datasource/pubsub/google/google_test.go:68]: assert.NoError please

File: pkg/gofr/datasource/pubsub/google/google_test.go [hunk 5]
```diff
@@ -50,6 +50,24 @@ func TestGoogleClient_New_Error(t *testing.T) {
...
+		g *googleClient
+	)
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	logger := logging.NewMockLogger(logging.ERROR)
+	g = New(Config{}, logger, NewMockMetrics(ctrl))
+
+	assert.Nil(t, g)
```
Comment (ccoVeille) [inline pkg/gofr/datasource/pubsub/google/google_test.go:64]: I don't get the `assert.Nil` , and then a `g.Close()` on next lines. It would lead to a nil pointer exception no?

File: pkg/gofr/container/container.go [hunk 6]
```diff
@@ -39,6 +40,21 @@ type Container struct {
...
 
+func (c *Container) Close() (err error) {
+	if c.PubSub != nil {
+		err = errors.Join(c.PubSub.Close())
+	}
+
+	if c.SQL != nil {
+		sqlDB, ok := c.SQL.(*sql.DB)
+		if ok && sqlDB != nil {
+			err = errors.Join(sqlDB.Close())
```
Comment (ccoVeille) [inline pkg/gofr/container/container.go:51]: Joining one error sounds odd

Either you wanted to join with another error, or you could use err =

File: pkg/gofr/container/container.go [hunk 7]
```diff
@@ -39,6 +40,21 @@ type Container struct {
 	Mongo datasource.Mongo
 }
 
+func (c *Container) Close() (err error) {
+	if c.PubSub != nil {
+		err = errors.Join(c.PubSub.Close())
```
Comment (ccoVeille) [inline pkg/gofr/container/container.go:45]: Joining one error sounds odd

Either you wanted to join with another error, or you could use err =

File: examples/http-server/main.go [hunk 8]
```diff
@@ -27,7 +31,12 @@ func main() {
 	a.GET("/mysql", MysqlHandler)
 
 	// Run the application
-	a.Run()
+	go a.Run()
+
+	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
```
Comment (ccoVeille) [inline examples/http-server/main.go:36]: I would have expect that this part would be in a subroutine and before the app is launched.

So the app would be without subroutine

File: pkg/gofr/gofr.go [hunk 9]
```diff
@@ -177,6 +178,29 @@ func (a *App) Run() {
 	wg.Wait()
 }
 
+// Shutdown stops the service(s) and close the application.
+func (a *App) Shutdown(ctx context.Context) (err error) {
```
Comment (ccoVeille) [inline pkg/gofr/gofr.go:182]: It seems to me you didn't implement a timeout processing as discussed in issue

https://github.com/gofr-dev/gofr/issues/663#issuecomment-2135108278

Maybe you would in another PR ?

File: examples/http-server/main.go [hunk 10]
```diff
@@ -27,7 +31,12 @@ func main() {
 	a.GET("/mysql", MysqlHandler)
 
 	// Run the application
-	a.Run()
+	go a.Run()
```
Comment (ccoVeille) [inline examples/http-server/main.go:34]: I would have refactored to pass the context to `a.Run` so the context could be used when calling data sources, and not only in the shutdown method

But that would be a huge refactoring that may change framework signature. Maybe it's what you were afraid of.

Maybe to be addressed in another PR/issue

**Review by ccoVeille — CHANGES_REQUESTED** [review]

**Comment by aryanmehrotra** [comment]
@maohieng are you working on this?

File: examples/http-server/main.go (hunk 10 above)
Comment (iamgoroot) [reply examples/http-server/main.go:34]: Probably signal handling and cancellation could be done inside `app.Run()` 
It would keep users from having to know that they have to implement handling shutdown by themselves, thus simplified examples and no reason to explain it in docs

**Comment by Umang01-hash** [comment]
Closing this PR due to inactivity from @maohieng.
